File: src/omnia/data/iterate/FilterIterator.java
package omnia.data.iterate;

import static java.util.Objects.requireNonNull;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.function.Predicate;

public class FilterIterator<E> implements Iterator<E> {
  private final Iterator<E> source;
  private final Predicate<? super E> filter;
  private boolean didPeekNext;
  private E peekedNext;
  private boolean hasNext;

  public FilterIterator(Iterator<E> source, Predicate<? super E> filter) {
    this.source = requireNonNull(source);
    this.filter = requireNonNull(filter);
  }

  @Override
  public boolean hasNext() {
    maybePeekNext();
    return hasNext;
  }

  @Override
  public E next() {
    maybePeekNext();
    if (!hasNext) {
      throw new NoSuchElementException("invoked next() when there is no next");
    }
    return clearNext();
  }

  private void maybePeekNext() {
    if (didPeekNext) {
      return;
    }
    didPeekNext = true;
    hasNext = false;
    while (source.hasNext()) {
      peekedNext = source.next();
      if (filter.test(peekedNext)) {
        hasNext = true;
        break;
      }
    }
  }

  private E clearNext() {
    didPeekNext = false;
    E next = peekedNext;
    peekedNext = null;
    return next;
  }

  @Override
  public void remove() {
    if (didPeekNext) {
      throw new IllegalStateException(
          "Cannot invoke remove() after calling hasNext(). This is a limitation of "
          + "FilterIterator.");
    }
    source.remove();
  }
}

File: src/omnia/data/iterate/FilterIterator.kt
package omnia.data.iterate

import java.util.NoSuchElementException
import java.util.function.Predicate

class FilterIterator<E>(private val source: Iterator<E>, private val filter: Predicate<in E>) :
  Iterator<E> {

  private var didPeekNext = false
  private var peekedNext: E? = null
  private var hasNext = false

  override fun hasNext(): Boolean {
    maybePeekNext()
    return hasNext
  }

  override fun next(): E {
    maybePeekNext()
    if (!hasNext) {
      throw NoSuchElementException("invoked next() when there is no next")
    }
    return clearNext()
  }

  private fun maybePeekNext() {
    if (didPeekNext) {
      return
    }
    didPeekNext = true
    hasNext = false
    while (source.hasNext()) {
      val next = source.next()
      peekedNext = next
      if (filter.test(next)) {
        hasNext = true
        break
      }
    }
  }

  private fun clearNext(): E {
    didPeekNext = false
    val next = peekedNext!!
    peekedNext = null
    return next
  }
}